src/pages/workspaces/workspace/useWorkspace.ts

```diff
@@ -256,14 +259,22 @@ export const useWorkspace = (namespace, name): WorkspaceDetails => {
     return withErrorIgnoring('Error loading workspace', doWorkspaceRefresh) as Promise<void>;
   };
 
-  useOnMount(() => {
+  useEffect(() => {
     if (!workspace) {
       refreshWorkspace();
+    } else if (workspace.workspace.namespace !== namespace || workspace.workspace.name !== name) {
+      // if the workspace is missmatched, clear store before fetching workspace
+      workspaceStore.reset();
+      refreshWorkspace();
     } else {
       checkWorkspaceInitialization(workspace);
     }
-    return () => clearTimeout(checkInitializationTimeout.current);
-  });
+    return () => {
+      abort();
+      clearTimeout(checkInitializationTimeout.current);
+    };
+    //  eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [name, namespace]);
```

Comment:
These are the correct dependencies, because we want this to re-load if the name/namespace changes; if the effect runs on every workspace change, it would trigger a loop when the workspace is set from this hook.

Reply:
But this is created with a specific name and namespace, and those can't be changed. So shouldn't this hook actually have NO dependencies? How could we ever end up going through this code more than once?

Reply:
On a re-render, react will re-run the hook based on the values of the deps array:

If you don't provide a deps array, the effect will be run on every re-render, which we don't want since it could be re-rendering for changes elsewhere in the tree that don't matter to this component, and we could therefore trigger extraneous API calls.

If you pass an empty array, which is what the `useOnMount` hook does, the effect will _never_ be run on a re-render. Normally a props change causes a re-render, not for the component to remount. So even if the props change, you wouldn't get the `useEffect` executing again, unless the component is remounted.

While we are often remounting the entire component on state changes, this behavior is usually an anti-pattern in react, and we shouldn't depend on it. 
Also, I was able to get the component to re-render with a different name/namespace in some circumstances, without remounting the component. I think it was by changing the url directly, but there may have been other ways it could happen as well.

From a standpoint of desired behavior, if the name/namespace changes, we want this effect to execute.
If the component is unmounted and remounted, the effect only executes on the remount. So even if we only had the unmount/remount behavior, adding the correct deps would not cause any issues.

Reply:
@nawatts I verified that if I remove `name` and `namespace` as dependencies, this effect still runs every time you switch tabs within a workspace or alter the workspace name to a different workspace (in the URL).

Reply:
Thanks for the explanation, @blakery. I agree that having them in the dependencies is the right thing to do… I was just struggling to understand how they can change within a re-render of a workspace since the name and namespace are not mutable.

Reply:
In what I'm convinced must have been a workaround to deal with the problems caused by `useOnMount`, Router sets the key of its child to the pathname. Thus, any navigation unmounts and recreates basically the entire component tree.

https://github.com/DataBiosphere/terra-ui/blob/bfddd4eab6195d85b5db58534e786603879979e4/src/libs/nav.ts#L136-L144

Reply:
Yes, I have noticed that switching between tabs in WorkspaceContainer causes _all_ the ajax calls to fire again… even for things that won't be used on that specific tab (like SAS token). It seems like something we really should fix (as a follow-up to this PR).